### PR TITLE
OAuth: persisting the id token

### PR DESCRIPTION
**Pull request by xlson**
**What this PR does / why we need it**:

This stores the oauth id token in the database which should fix #42422. This is a draft PR as I still need to configure oauth to test this and I'm not sure I will be able to do that today but wanted to share this now.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

@pascaldekloe Could you give this a try? It's untested but I'm going to set up an oauth config myself to test it but thought you might want to take a look right away as well as well.

### CODE REVIEW

**Comment by pascaldekloe** [comment]
Yes this is [my patch](https://github.com/grafana/grafana/pull/41837#issuecomment-981666591) and it works.

The database migration added in this PR may have an issue. I think we can't just add the column with a zero value. It should be safe to delete all existing records (in the same commit) though.

Again, I'm happy to write the whole thing, as long as you express which solution is preferred. The `user_auth_mig.go` file option looks good.

File: pkg/services/sqlstore/migrations/user_auth_mig.go [hunk 1]
```diff
@@ -42,4 +42,8 @@ func addUserAuthMigrations(mg *Migrator) {
 	mg.AddMigration("Add index to user_id column in user_auth", NewAddIndexMigration(userAuthV1, &Index{
 		Cols: []string{"user_id"},
 	}))
+
+	mg.AddMigration("Add OAuth ID token to user_auth", NewAddColumnMigration(userAuthV1, &Column{
+		Name: "o_auth_id_token", Type: DB_Text, Nullable: true,
```
Comment (pascaldekloe) [inline pkg/services/sqlstore/migrations/user_auth_mig.go:47]: No need for `null`. The empty string works well here.

**Comment by xlson** [comment]
Glad to hear that the fix is working for you.

> The database migration added in this PR may have an issue. I think we can't just add the column with a zero value. It should be safe to delete all existing records (in the same commit) though.

Do you mean that we should delete all auth info records as a part of the migration? We can't really do that since that would sign out all users using oauth when upgrading, I don't think this change warrants that behavior.

But it could be that we need an empty string to be the default value instead of `null`, I'll look into that.

File: pkg/services/sqlstore/migrations/user_auth_mig.go (hunk 1 above)
Comment (xlson) [reply pkg/services/sqlstore/migrations/user_auth_mig.go:47]: Sticking to `null` as the default value as that is what is being used for the other fields, no point in changing the convention.

**Comment by pascaldekloe** [comment]
Always in character @xlson. 🤣 Sure go ahead.

**Review by sakjur — APPROVED** [review]
:+1:

For future reference: This doesn't support `id_token_attribute_name`, I've discussed that with @xlson and we're in agreement that that's not a prerequisite for this PR to be merged.

**Comment by pascaldekloe** [comment]
Indeed @sakjur. The generic way to persist a token would be a JSON of the entire "extra" map.

**Pull request merged**